### ISSUE TITLE
Clarify @LoginConfig usage

### ISSUE THREAD

**Issue by jeanouii** [eclipse/microprofile-jwt-auth#70]
I have been implementing MP-JWT in TomEE.
And it's unclear to me where to use @LoginConfig.

I have assumed it can only be used to qualify a JAX RS Application.
It seems that Swarm did the same call (not sure though).

I see no problem technically to use @LoginConfig on a single JAX RS Resource.

What do you think?

**Comment by rdebusscher** [comment]
There are more things unclear about @LoginConfig usage.

According to the TCK tests (see UnsecuredPingTest), even when we have a JAX-RS application with @LoginConfig, a request matching the URL doesn't need to have the Authorization header with a valid JWT bearer token present.
In that case, no specific security context is setup and endpoints with no security requirements (like @RolesAllowed is not present) still can be executed.
It is not the same situation as a BASIC authentication mechanism where credentials need to be present all the time.

**Comment by rdebusscher** [comment]
an extended @LoginConfig section in the spec should have something like

- can it be placed on a JAX-RS resource and thus it affects then all Resources in that JAX RS application? (when JAX-RS Application class present or not)
- @LoginConfig doesn't mandate that JWT Bearer token needs to be present. But if presented, it should be used to set up the security context.
- What is the behavior when authMethod is not MP-JWT

**Comment by arjantijms** [comment]
My proposal would be to deprecate `@LogingConfig`. It hints at being able to set other authentication mechanisms (such as FORM), but that's not a concern that the JWT spec should address.

JWT should only offer an annotation to state that the JWT authentication mechanism is to be used. 

My proposal:

```java
@JWTAuthenticationMechanismDefinition(
    publicKey = “#{microprofile.config[some.config.key]}”
    acceptedIssuers = #{microprofile.config[some.accepted_issues]}

    // other standard attributes here, e.g. preEmptive, clock skew, etc 
}
```

**Comment by rdebusscher** [comment]
The @JWTAuthenticationMechanismDefinition is the style of Java EE Security/Soteria, but I have the impression that MP uses another style

- Shorter names (like RestClient)
- Config not as members but just read from config (like com.mycompany.remoteServices.MyServiceClient/mp-rest/url)

Although I like the way Java EE Security/Soteria style, we should stay consistent.

=> Something for the architecture board?

**Comment by arjantijms** [comment]
Hi,

*definition is also the style of Java EE in general, for example
@MailSessionDefinition, @DataSourceDefinition, etc.

As MP is Java EE based, and there's talk about potentially merging them
into Java EE/Jakarta EE at some point, it's probably a good idea to be
consistent with that style.

MP-JWT already derived from that config by introducing the @LoginConfig
annotation for saying that you want JWT Authentication in the first place.
This proposal is effectively just renaming the annotation and specifying
that it's for JWT only, without any pretending it's only usable for an
open-ended number of other mechanisms. This is basically what all
implementations use it for today in practice anyway.

The new thing is putting the config in the annotation as attributes, but
even that isn't totally new as the @LogingConfig annotation already had one
attribute used for config (but this one seems to be ill understood, as the
docs say it's only used for BASIC, but JBoss uses it for their security
domain concept, and I think also to put the key and issuer in as an
optional way. E.g.

@LoginConfig(
    authMethod="MP=JWT",
    realmName="JBOSS-SECURITY-DOMAIN?issuer=myIssuer"
)

The advantage of annotation attributes here is that they're
auto discoverable and type safe for their names, plus by being EL enabled
they can still easily use MP Config.

The name could be shortened still, e.g.

@JWTConfig(
    issuer = "#{microprofile.config['my.accepted.issuer']}"
)

That's even 2 characters shorter than the current @LoginConfig ;)

Kind regards,
Arjan










On Fri, Mar 23, 2018 at 12:08 PM, Rudy De Busscher <notifications@github.com

**Comment by arjantijms** [comment]
Hi,

*definition is also the style of Java EE in general, for example
@MailSessionDefinition, @DataSourceDefinition, etc.

As MP is Java EE based, and there's talk about potentially merging them
into Java EE/Jakarta EE at some point, it's probably a good idea to be
consistent with that style.

MP-JWT already derived from that config by introducing the @LoginConfig
annotation for saying that you want JWT Authentication in the first place.
This proposal is effectively just renaming the annotation and specifying
that it's for JWT only, without any pretending it's only usable for an
open-ended number of other mechanisms. This is basically what all
implementations use it for today in practice anyway.

The new thing is putting the config in the annotation as attributes, but
even that isn't totally new as the @LogingConfig annotation already had one
attribute used for config (but this one seems to be ill understood, as the
docs say it's only used for BASIC, but JBoss uses it for their security
domain concept, and I think also to put the key and issuer in as an
optional way. E.g.

```java
@LoginConfig(
    authMethod="MP=JWT",
    realmName="JBOSS-SECURITY-DOMAIN?issuer=myIssuer"
)
```

The advantage of annotation attributes here is that they're
auto discoverable and type safe for their names, plus by being EL enabled
they can still easily use MP Config.

The name could be shortened still, e.g.

```java
@JWTConfig(
    issuer = "#{microprofile.config['my.accepted.issuer']}"
)
```

That's even 2 characters shorter than the current @LoginConfig ;)

**Comment by rdebusscher** [comment]
I agree that it is much more clear when you have additional attributes.
But MicroProfile isn't based on EL spec and thus not able to use EL Expressions.

**Comment by arjantijms** [comment]
True, but MicroProfile's base is not set in stone (or "wat niet is kan nog komen" ;)), and EL is easy enough to add. Even as an optional dependency; if EL isn't present then expressions can't be evaluated, but otherwise everything would just work.

**Comment by rmannibucau** [comment]
+1 to JL's proposal, @LoginConfig can be on any CDI bean technically and it makes sense
-1 to define a @JwtAuthenticationDefinition (based on the EE pattern), all the config is about environment configuration and must rely - for consistency - on MP-config so no real point allowing to hardcode it for now IMHO. Proposal also enforces some EL (or EL) dependency which is not the case today which is sane since most of environments will not have it or don't want to have it for that small need. Should be fully deferred to MP-config which can have a syntax evaluation probably.

**Comment by arjantijms** [comment]
>-1 to define a @jwtauthenticationdefinition (based on the EE pattern)

I think it's probably not bad to align with EE. MicroProfile should eventually be merged with Jakarta EE (as a profile of it) and already the idea is that MicroProfile APIs work together with EE APIs.

MicroProfile is not a completely different build from scratch ecosystem, but already heavily leans on EE. To willingly go into different directions where there's no need makes little sense to me.

`@LoginConfig` has been rather hastily added, since there was a desire for JWT support, but MP actually didn't had any security backbone. The annotation as-is makes little sense. It has a `realm` attribute that's only for BASIC authentication, but nobody uses it for BASIC authentication and `realm` is used by JBoss for 'protection domain`. The JavaDoc says that it may support FORM as well, but where are the `loginPage` and `errorPage` attributes?

A  `@JWTConfig` or `@JwtAuthenticationDefinition` annotation would not have any ties to EE security. And it doesn't HAVE to support EL in its attributes, but it can if the environment supports it.

The biggest problem is that we're repeating history again here. JWT does it in a specific style, but in 3 months time (say), OAuth will be done in EE Security and it will use a different style. Then developers using it will again have to deal with an inconsistent platform, just like is currently the case with JAX-RS annotations and CDI annotations etc.

**Comment by rmannibucau** [comment]
Ok, so I join you on the fact @LoginConfig is useless. What about just @Deprecated it and just not use it at all? If the server is configured to use jwt-auth then set it up, else ignore it. That's enough and solves all issues.

**Comment by cghislai** [comment]
I second the drop of LoginConfig and the addition of a JWT config annotation, similar to the others already present in the security api. Configuration could be set/overriden using the mp-config api, similar to what they did for the fault tolerance annotations (https://github.com/eclipse/microprofile-fault-tolerance/blob/master/spec/src/main/asciidoc/configuration.asciidoc)

**Comment by sberyozkin** [comment]
I would be cautious about deprecating LoginConfig. 
The reason is that the only login type which is supported now is MP-JWT. This does not layer with the existing deployments where the tokens are issued by OpenId Connect providers and the JAX-RS endpoint RPs are already protected by those provider adapters with the provider-specific configurations. 
For example, if Keycloak issues a token, it will be validated by a Keycloak specific RP adapter.
But I'd still like to access the Keycloak-issued tokens using MP-JWT API which will give me a portable code.
In many cases like this one the existing 3rd party RP adapter configuration would not be portable to the MP-JWT configuration.
But what would be possible is to get Keycloak (or other) adapter validate the token and then feed in the results into MP-JWT. This would require MP-JWT setting a different login type, for ex, "KEYCLOAK" (as well as plugging in a diff validation mechanism - in this case it will just collect Keycloak validation results, etc)
So IMHO the option to specify the alternative login type should be open. It is important to be able to use MP-JWT to interpose on top of existing SSO deployments which use JWT tokens

**Comment by rmannibucau** [comment]
Hmm, is it right? I mean security mecanism will wrap themselves (you can see it as the plain old servlet filter chain but at principal level) so the one able to handle it will and there is still no need to have @LoginConfig on user land, no?

**Comment by sberyozkin** [comment]
Romain, I don't understand what exactly are you questioning in my comment. I have tried to advocate that activating MP JWT with a non-default login type value should be possible for the reasons outlined above. IMHO it will make MP JWT more deployable. 

Typical line could be "if you switch from a Keycloak/etc specific code to access the token in your service code to MP-JWT API then it will make your service code portable" which is realistic to implement as opposed to "Replace all your Keycloak (etc) validation configuration with MP JWT one" - because MP JWT won't able to map all the 3rd party OIDC client adapter configurations from various vendors.

I thought LoginConfig was the way to be able to set the alternative login type. However I also agree with @arjantijms that there is a bit of ambiguity to do with the login types which have nothing to do with JWT. So may be the spec can restrict that some well-known login types can ot be set...

**Comment by sberyozkin** [comment]
@rmannibucau Have you really read what I've suggested ? I would like to access the tokens coming from Keycloak into my (JAX-RS) service code with MP-JWT API as opposed with Keycloak specific API. I don't need MP-JWT to activate Keycloak, I only need it to ensure the token access code is portable

**Comment by sberyozkin** [comment]
@rmannibucau Though may be it is me who did not read your note carefully :-). However, it might not be realistic to implement what you are suggesting - by the time the servlet filter is hit it should already be secured.

**Comment by rmannibucau** [comment]
@sberyozkin hmm, the security is done through a filter - or equivalent like a valve in tomcat or so - so you can always combine the MP-JWT one with another one to make what you describe with keycloak, i see no blocker to have it already, even without this annotation. This is even better cause it increases the decoupling between the security and the business code (which is not the case with this annotation)

**Comment by sberyozkin** [comment]
Romain, many thanks for the feedback, I thought about what you said... I'll need to experiment, cheers

**Comment by rmannibucau** [comment]
Doesnt deactivating the validations and implementing in a filter already works?

Will avoid to enrich mp-config genericly (dont think it will hit an annotation cause it is deployment related so config specific).

Alternative is to have a spi but it makes the lib quickly overcomplicated and therefore not used and reimplemented on need since this part is not crazy to do so personally i prefer to ensure the composability than extending and abstracting what exists.

**Comment by sberyozkin** [comment]
I'm not sure I follow your line. It does not matter how a given MP-JWT implementation is done. But, it is important to recognize the fact that the existing configuration support is geared toward a 'local' token validation - which will work in some cases but will not in others, hence it would be good to support at the MP JWT level that this is one one, local, validation mechanism.
As I said, one other obvious alternative is to use [RFC7622](https://tools.ietf.org/html/rfc7662) to validate the tokens issued by OIDC providers, and any decent OIDC provider does support RFC7622, hey, even Apache Fediz supports it :-). So if one were to validate it with RFC7622 then all the current configuration for validating locally would not apply.
The other obvious alternative, is, as I tried to articulate at length above is to use the 3rd party configuration (and I can expand more on it).
To summarize, IMHO, realizing at the spec level that different validation techniques require different current validation configuration would be of great benefit to all the MP JWT implementers

**Comment by rmannibucau** [comment]
The point is that any toggable abstraction will likely make the spec complicated enough to use to be tempted to be bypassed by users. Since you can already add any validation and support any token - and multiple auth types, im not sure it is worth to take that risk *at spec level*, vendors can propose it and it would bring some user base to revise that in a year or so. All that does not need @LoginConfig at all so this annotation should be dropped.

What can be worth is to standardize an optinal servlet filter to integrate with user filters this feature (filter chain) cause jaxrs filters are far to be mainstream yet but this is more about the ecosystem than the API.

**Comment by sberyozkin** [comment]
I'm afraid you are missing the point I'm trying to make. The existing configuration can not handle various validation scenarios (I can expand again why - but I am not sure you will read) and servlet filters you are referring to is nothing to do with what I'm trying to say either.

**Comment by sberyozkin** [comment]
Only one more brief clarification re LoginConifg. Basically, I'm following Arjan's lead but I'm only suggesting to have one annotation that will activate MP-JWT subsystem, and another one - related to the validation configuration - that will easily enable the future versions of the specs to 1) support different validation requirements (ex - RFC7622) and other alternatives (ex 3rd party OIDC configuration can not be migrated to MP-JWT, etc)

**Comment by sberyozkin** [comment]
Oh, well :-), one more comment, sorry all. Romain, you said above " Since you can already add any validation and support any token"... - this is just not true that it can be done with the MP-JWT validation configuration and I'd challenge you to prove it otherwise (I'll be happy to offer you some test ideas).

**Comment by rdebusscher** [comment]
I do not want to interfere with this discussion (as there is probably a good use case for multiple types of validation of the token), but basically, it comes to this (the discussion of the last 2 weeks o this issue)

Do we base ourself on Servlets or JAX-RS/CDI?

In light of that question, the answer is simple. We don't have servlets in MicroProfile. It was already discussed in the MP-JWT weekly hangouts and at the MP Architecture meetings.

**Comment by rmannibucau** [comment]
@rdebusscher yes and no, if not integrated with servlet the servlet part of the TCK can't work portably on user side so the impl must be servlet related at the end - until you want to duplicate integrations without any gain. That said the same applies to JAX-RS filters so the point is exactly the same.

**Comment by sberyozkin** [comment]
Hi Romain, of course you are right one can do it at the filter level (or at a pluggable validation factory leve) - this is something you clearly told me about earlier and yes, it works. But it is not what I'm after now :-). I spotted Arjan's proposal which introduces the annotation which has 2 aspects in it 1) MP-JWT activation and 2) MP-JWT validation configuration. So if such an annotation were to be introduced then it would be impossible to introduce at the spec level a *portable* support for the alternative validation techniques => I proposed to make a bit more flexible.

As for the test case, a token encrypted at the OIDC provider level would be impossible to decrypt at the MP-JWT RS level. Well technically it is possible but not practically because OIDC server private keys would have to be distributed to every MP JWT node or some complex key management protocol involved... So one could just simply use RFC 7622 instead in such a case. Or another case, OIDC providers can for ex restrict a number of times the access token is used, esp for long-lived ones - possibly on a client-specific base. Etc. 
MP-JWT covers a good number of cases given that most tokens today are JWS-signed tokens, it is good. But anticipating more sophisticated validation scenarios will serve MP-JWT well :-)
In any case, thanks for the feedback, perhaps we can indeed do some shared MP-JWT validation project and collaborate again :-), and happy Christmas

**Comment by rmannibucau** [comment]
Yep. This issue is really about 1 and I really think we can drop the annotation, it doesn't bring anything. Now 2 is a real question, do we enable a default validator and put a config to switch it to something else - strictly speaking it is more a config than an annotation - at least overridable - cause it can depend where you deploy your app.

**Comment by sberyozkin** [comment]
Hi All, I'm going to close this issue for the following reasons:
* too many different issues have been discussed here and even I managed to have a good discussion with Romain, as we have always had before :-)
* IMHO, doing anything with `LoginConfig` is not worth spending the time/energy upon. Please don't take me wrong, it is not perfect, but it works for all.
* I'm against narrowing the scope of `LoginConfig` to anything lower than JAX-RS `Application` because: I'm not a fan of `PermitAll` for a start, as I don't think that any application serious enough about its security will allow anyone in the WEB hitting one of its methods while ultra secured methods sit in the same class protected by `RolesAllowed`. Exactly for the same reason I'm against have an `Application` with N resource classes, where only one of them has `LoginConfig` (or something else instead of it) while N-1 resources are open.

Please appreciate I'm closing the issues also from my Red Hat colleagues - the goal is to have a manageable set of issues which can be tackled independently, without any confusion. If you look at the issue list it is now quite concrete.

As I've said elsewhere, if you have any concerns, disagree - then just please open the very specific issues you may have touched somewhere in this discussion :-). @arjantijms - may be you can update the EE security issue with some more suggestions how to deal with LoginConfig or something better in the EE security scope, etc

Thanks All

**Comment by arjantijms** [comment]
> may be you can update the EE security issue with some more suggestions how to deal with LoginConfig

This is essentially already there, `LoginConfig` just specifies that the JWT-AUTH authentication mechanism should be used. I implemented JWT-AUTH myself using EE Security, and there it's treated exactly as that:

See https://github.com/omnifaces/jwt/blob/master/src/main/java/org/omnifaces/jwt/cdi/CdiExtension.java#L110

This is the same as annotations for other EE Security authentication mechanisms are being used. The only real difference is that `@LoginConfig` has a different naming convention.
